### PR TITLE
Collect multiple paths from XContentMapValues

### LINKED ISSUE #68215: Strange behaviour of `fields` API with mixed object notation
It appears we can currently index and retrieve documents with mixed object notation like this:
```
PUT my-index-000001/_doc/1
{
  "obj.products" : { "manufacturer" : "Supersoft" } ,
  "obj" : {
    "products" : 
       { "manufacturer" : "HyperSmart" }
  }
}
```

When searching on field `obj.products.manufacturer` we get a hit for both terms "Supersoft" and "HyperSmart". 
However, the `fields` API output only returns second value:
```
POST /my-index-000001/_search
{
  "query": {
            "match": {
              "obj.products.manufacturer": "Supersoft"
            }
    },
  "_source" : false,
  "fields" : ["*"]
}
```
returns
```
       "fields" : {
          "obj.products.manufacturer.keyword" : [
            "HyperSmart"
          ],
          "obj.products.manufacturer" : [
            "HyperSmart"
          ]
        }
```
The same output is produced when searching for "HyperSmart". Tested on 7.10.0.

### PR DESCRIPTION
**Pull request by cbuescher**
Currently, when a document source mixed json object and dotted syntax like e.g.
{ "foo" : { "bar" : 0 }, "foo.bar" : 1}, extracting the values from the source
map via XContentMapValues#extractValue returns after the first value for a path
has been found. Instead we should exhaust all possibilities and return a list of
objects of we find more than one value when extending the lookup path.

Closes #68215

### CODE REVIEW

**Comment by cbuescher** [comment]
I think `extractRawValues` would use a similar treatment but wanted to ask about the differences in these two methods before going ahead and adding it. Can do so in this PR though when I understand the expectations better.

File: server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java [hunk 1]
```diff
@@ -195,6 +196,17 @@ public void testExtractValueMixedObjects() throws IOException {
...
 
+    public void testExtractValueMixedDottedObjectNotation() throws IOException {
+        XContentBuilder builder = XContentFactory.jsonBuilder().startObject()
+            .startObject("foo").field("cat", "meow").endObject()
+            .field("foo.cat", "miau")
+            .endObject();
+        try (XContentParser parser = createParser(JsonXContent.jsonXContent, Strings.toString(builder))) {
+            Map<String, Object> map = parser.map();
+            assertThat((List<?>) XContentMapValues.extractValue("foo.cat", map), containsInAnyOrder("meow", "miau"));
+        }
```
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java:207]: What happens if we have `"foo.cat" : ["miaw", "purr"]` here?  I guess we get a List that contains a String and a further List, which is probably fine given that this is an edge case already.

**Review by romseygeek — APPROVED** [review]
LGTM, thanks @cbuescher

File: server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java (hunk 1 above)
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java:207]: Good point. I think flattening out the list in this case makes sense and is doable. In the end we only need to handle single values and lists here. I added a commit along those lines and extended the test.

**Comment by cbuescher** [comment]
@romseygeek thanks for the review, what do you think about my question above whether this addition should also happen in `extractRawValues`. I don't fully understand the different scenarios in which those two variations are used, thats why I left it out for now.

File: server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java (hunk 1 above)
Comment (romseygeek) [reply server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java:207]: This looks like it breaks a bunch of assumptions in tests - let's leave it as it was previously, with the mix of values? As I said, I think it's a sufficiently rare case that having an odd-looking response in fetch is fine.

**Comment by romseygeek** [comment]
> I think extractRawValues would use a similar treatment but wanted to ask about the differences in these two methods before going ahead and adding i

`extractRawValues` already has this behaviour (see #65539)

File: server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java (hunk 1 above)
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java:207]: Saw the tests, probably just a class cast issue that I overlooked. Will see how easy it is to fix

Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java:207]: Okay, I can see how this gets tricky now, rolled back the last change. I added a test to the FieldFetcherTests though that shows that although we might return a List that contains a String and a further List in this case via XContentMapValues (which can even be considered to be correct, depending on the viewpoint), we unpack this e.g. when retrieving fields values to a simple list.

**Review by romseygeek — APPROVED** [review]
LGTM! Thanks for the iterations @cbuescher

**Comment by cbuescher** [comment]
@elasticmachine update branch

**Comment by cbuescher** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by cbuescher** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by cbuescher** [comment]
@elasticmachine update branch

**Pull request merged**